src/backend/mod.rs

```diff
@@ -144,67 +135,132 @@ pub fn wait_for_gdb_connection(sockaddr: &str) -> std::io::Result<std::net::TcpS
-                    .map(|_| [0usize.into(), 0usize.into()])
+                    .map(|_| 0usize)
                     .map_err(|e| e.raw_os_error().unwrap_or(libc::EINVAL));
+
                 if res.is_ok() {
                     gdb_fd.replace(stream);
                 }
-                block.msg.rep = res.into();
+
+                *ret = match res {
+                    Ok(n) => n as usize,
+                    Err(e) => -e as usize,
+                };
             } else {
-                block.msg.rep = Ok([0usize.into(), 0usize.into()]).into();
+                *ret = 0;
             }
+            Ok(())
         }
 
-        sallyport::syscall::SYS_ENARX_GDB_PEEK => {
+        item::Gdbcall {
+            num: item::gdbcall::Number::Flush,
+            ret,
+            ..
+        } => {
             let stream = gdb_fd.as_mut().unwrap();
 
-            let ret = Connection::peek(stream)
-                .map(|v| {
-                    let v = v.map(|v| v as usize).unwrap_or(u8::MAX as usize + 1);
-                    [v.into(), 0usize.into()]
-                })
+            let res = Connection::flush(stream)
+                .map(|_| 0)
                 .map_err(|e| e.raw_os_error().unwrap_or(libc::EINVAL));
-            block.msg.rep = ret.into();
+
+            *ret = match res {
+                Ok(n) => n as usize,
+                Err(e) => -e as usize,
+            };
+            Ok(())
         }
 
-        sallyport::syscall::SYS_ENARX_GDB_READ => {
+        item::Gdbcall {
+            num: item::gdbcall::Number::Peek,
+            ret,
+            ..
+        } => {
             let stream = gdb_fd.as_mut().unwrap();
 
-            let buf_ptr: *mut u8 = req.arg[0].into();
-            let buf_len: usize = req.arg[1].into();
-            let buf = unsafe { core::slice::from_raw_parts_mut(buf_ptr, buf_len) };
+            let res = Connection::peek(stream)
+                .map(|v| v.map(|v| v as usize).unwrap_or(u8::MAX as usize + 1))
+                .map_err(|e| e.raw_os_error().unwrap_or(libc::EINVAL));
+            *ret = match res {
+                Ok(n) => n,
+                Err(e) => -e as usize,
+            };
+            Ok(())
+        }
+
+        item::Gdbcall {
+            num: item::gdbcall::Number::Read,
+            ret,
+            ..
+        } => {
+            let stream = gdb_fd.as_mut().unwrap();
 
-            let ret = stream
-                .read_exact(buf)
-                .map(|_| [buf_len.into(), 0usize.into()])
+            let res = stream
+                .read()
                 .map_err(|e| e.raw_os_error().unwrap_or(libc::EINVAL));
 
-            block.msg.rep = ret.into();
+            *ret = match res {
+                Ok(n) => n as usize,
+                Err(e) => -e as usize,
+            };
+            Ok(())
         }
 
-        sallyport::syscall::SYS_ENARX_GDB_WRITE => {
+        item::Gdbcall {
+            num: item::gdbcall::Number::Write,
+            argv,
```

Comment:
Why not pattern match on this to avoid indexing and for consistency?

Reply:
thx